codegenerator/cli/templates/static/codegen/src/eventFetching/SourceManager.res

```diff
@@ -240,3 +252,103 @@ let waitForNewBlock = async (sourceManager: t, ~currentBlockHeight) => {
 
   newBlockHeight
 }
+
+let executeQuery = async (sourceManager: t, ~query: FetchState.query, ~currentBlockHeight) => {
+  let toBlock = switch query.target {
+  | Head => None
+  | EndBlock({toBlock})
+  | Merge({toBlock}) =>
+    Some(toBlock)
+  }
+
+  let allAddresses = query.contractAddressMapping->ContractAddressingMap.getAllAddresses
+  let addresses =
+    allAddresses->Js.Array2.slice(~start=0, ~end_=3)->Array.map(addr => addr->Address.toString)
+  let restCount = allAddresses->Array.length - addresses->Array.length
+  if restCount > 0 {
+    addresses->Js.Array2.push(`... and ${restCount->Int.toString} more`)->ignore
+  }
+
+  let responseRef = ref(None)
+
+  while responseRef.contents->Option.isNone {
+    let source = sourceManager.activeSource
+
+    let logger = Logging.createChild(
+      ~params={
+        "chainId": source.chain->ChainMap.Chain.toChainId,
+        "logType": "Block Range Query",
+        "partitionId": query.partitionId,
+        "source": source.name,
+        "fromBlock": query.fromBlock,
+        "toBlock": toBlock,
+        "addresses": addresses,
+      },
+    )
+
+    try {
+      let response = await source.getItemsOrThrow(
+        ~fromBlock=query.fromBlock,
+        ~toBlock,
+        ~contractAddressMapping=query.contractAddressMapping,
+        ~partitionId=query.partitionId,
+        ~currentBlockHeight,
+        ~selection=query.selection,
+        ~logger,
+      )
+      logger->Logging.childTrace({
+        "msg": "Fetched block range from server",
+        "latestFetchedBlockNumber": response.latestFetchedBlockNumber,
+        "numEvents": response.parsedQueueItems->Array.length,
+        "stats": response.stats,
+      })
+      responseRef := Some(response)
+    } catch {
+    | Source.GetItemsError(error) => {
+        // TODO: When we start handling fetch failures,
+        // we shouldn't delete the source from the set
+        // but keep it for retries. Still need to delete
+        // for cases like UnsupportedSelection which are
+        // not retryable
+        let notAlreadyDeleted = sourceManager.sources->Utils.Set.delete(source)
+
+        // In case there are multiple partitions
+        // failing at the same time. Log only once
+        if notAlreadyDeleted {
+          switch error {
+          | UnsupportedSelection({message}) => logger->Logging.childError(message)
+          | FailedGettingFieldSelection({message, blockNumber, logIndex})
+          | FailedParsingItems({message, blockNumber, logIndex}) =>
+            logger->Logging.childError({
+              "msg": message,
+              "blockNumber": blockNumber,
+              "logIndex": logIndex,
+            })
+          }
+        }
+
+        switch sourceManager->getNextActiveSource {
+        | None => {
```

Comment:
This is not the most robust logic, and can be improved in the future. But currently it only applies to RPC data-sources and good enough to prevent Fallback sources from killing the indexer when there are still other sources.